### PR TITLE
Add: util/atomic.h for Cortex CPU

### PR DESCRIPTION
**Pull request by d00616**
The atomic.h file I have taken from Teensy and tested with nRF51 and nRF52.

### CODE REVIEW

File: cores/nRF5/Print.cpp [hunk 1]
```diff
@@ -37,6 +37,16 @@ size_t Print::write(const uint8_t *buffer, size_t size)
...
+// for printf compatibility
+extern "C" {
+__attribute__((weak))
+int _write(int file, char *buffer, int size)
+{
+  Serial.write(buffer, size);
+	return 0;
+}
+}
+
```
Comment (sandeepmistry) [inline cores/nRF5/Print.cpp]: This shouldn't be here?

File: cores/nRF5/util/atomic.h [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+* >>I think it's great that you've ported the macros; consider them
+* >>public domain, to do with whatever you wish. I hope you find them >useful .
+* >>
+* >>Cheers!
+* >>- Dean 
+*/
+
+#ifdef __arm__ 
+#ifndef _CORTEX_M3_ATOMIC_H_
+#define _CORTEX_M3_ATOMIC_H_
```
Comment (sandeepmistry) [inline cores/nRF5/util/atomic.h]: Does this work on both nRF51 (M0) and nRF52 (M4)?

Comment (d00616) [inline cores/nRF5/util/atomic.h]: I have tested with both MCU. Here is my test sketch. The "counter" is increased circa once a second until 10 seconds. Then a 5 seconds delay is executed in an "ATOMIC" block.

```
#include "nrf.h"
#include <util/atomic.h>

uint16_t counter = 0;

void setup() {
  Serial.begin(38400);

  NVIC_SetPriority(RTC0_IRQn, 15);
  NVIC_ClearPendingIRQ(RTC0_IRQn);
  NVIC_EnableIRQ(RTC0_IRQn);
  NRF_RTC0->PRESCALER = 32;
  NRF_RTC0->EVTENSET = RTC_EVTENSET_COMPARE0_Msk;
  NRF_RTC0->INTENSET = RTC_INTENSET_COMPARE0_Msk;
  NRF_RTC0->CC[0] = 500;
  NRF_RTC0->TASKS_START = 1;
}

void loop() {
  ATOMIC_BLOCK(ATOMIC_FORCEON)
  {
     if (millis()>10000) {
      delay(5000);
     }
  }
  Serial.print("Counter: ");
  Serial.println(counter);
  Serial.print("RTC: ");
  Serial.println(NRF_RTC1->COUNTER);
  Serial.print("Millis: ");
  Serial.println(millis());
  delay(300);
}

extern "C" void RTC0_IRQHandler(void) { counter++; NRF_RTC0->EVENTS_COMPARE[0] = 0; NRF_RTC0->CC[0] = NRF_RTC0->COUNTER + 500; } 
```

File: cores/nRF5/Print.cpp (hunk 1 above)
Comment (d00616) [inline cores/nRF5/Print.cpp]: Sorry. I have removed this change.

**Comment by sandeepmistry** [comment]
@d00616 thanks for the info.

I'm still on the fence about including this. The header file is useful, however it's not really part of the standard Arduino API, so using it makes the code less portable.

**Comment by d00616** [comment]
@sandeepmistry This code comes per default with Arduino for AVR and Teensy. There are a lot of sketches using ATOMIC_BLOCK macro. Supporting it helps to make Sketches portable from AVR/Teensy to nRF5. There are other useful definitions in the "utils" folder comming with AVR arduino.

**Comment by sandeepmistry** [comment]
@d00616 yes, but it's not part of the official [SAMD](https://github.com/arduino/ArduinoCore-samd) and [SAM](https://github.com/arduino/Arduino/tree/master/hardware/arduino/sam) cores. Plus they are not documented in the [Arduino Reference docs](https://www.arduino.cc/en/Reference/HomePage).

So overall, only 1 out of 3 of the official Arduino cores have them (AVR), and it's only in AVR because it's bundled with the AVR toolchain.

**Comment by sandeepmistry** [comment]
Closing this for now, the SAMD core does not include `util/atomic.h`.

**Pull request closed without merging**